include/nbl/builtin/hlsl/math/functions.hlsl

```diff
@@ -0,0 +1,609 @@
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_HLSL_MATH_FUNCTIONS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_MATH_FUNCTIONS_INCLUDED_
+
+#include <nbl/builtin/hlsl/math/constants.hlsl>
+#include <nbl/builtin/hlsl/common.hlsl>
+
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace math
+{
+
+namespace impl
+{
+
+static float3 reflect_refract(in bool _refract, in float3 I, in float3 N, in float NdotI, in float NdotTorR, in float rcpOrientedEta)
+{    
+    return N*(NdotI*(_refract ? rcpOrientedEta:1.0)+NdotTorR) - I*(_refract ? rcpOrientedEta:1.0);
+}
+
+static uint bitfieldInsert(in uint base, in uint shifted_masked_value, in uint lo, in uint count)
+{
+    const uint hi = base^lo;
+    return (hi<<count)|shifted_masked_value|lo;
+}
+
+}
+
+
+int dot(in int2 a, in int2 b)    { return a.x*b.x+a.y*b.y; }
+uint dot(in uint2 a, in uint2 b) { return a.x*b.x+a.y*b.y; }
+int dot(in int3 a, in int3 b)    { return a.x*b.x+a.y*b.y+a.z*b.z; }
+uint dot(in uint3 a, in uint3 b) { return a.x*b.x+a.y*b.y+a.z*b.z; }
+int dot(in int4 a, in int4 b)    { return a.x*b.x+a.y*b.y+a.z*b.z+a.w*b.w; }
+uint dot(in uint4 a, in uint4 b) { return a.x*b.x+a.y*b.y+a.z*b.z+a.w*b.w; }
+
+
+template<typename vector_t>
+vector_t erf(in vector_t _x)
+{
+    const float a1 = 0.254829592;
+    const float a2 = -0.284496736;
+    const float a3 = 1.421413741;
+    const float a4 = -1.453152027;
+    const float a5 = 1.061405429;
+    const float p  = 0.3275911;
+
+    vector_t _sign = sign(_x);
+    vector_t x = abs(_x);
+    
+    vector_t t = 1.0 / (1.0 + p*x);
+    vector_t y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * exp(-x * x);
+    
+    return _sign*y;
+}
+
+
+
+float erfInv(in float _x)
+{
+    float x = clamp(_x, -0.99999, 0.99999);
+    float w = -log((1.0-x) * (1.0+x));
+    float p;
+    if (w<5.0)
+    {
+        w -= 2.5;
+        p = 2.81022636e-08;
+        p = 3.43273939e-07 + p*w;
+        p = -3.5233877e-06 + p*w;
+        p = -4.39150654e-06 + p*w;
+        p = 0.00021858087 + p*w;
+        p = -0.00125372503 + p*w;
+        p = -0.00417768164 + p*w;
+        p = 0.246640727 + p*w;
+        p = 1.50140941 + p*w;
+    }
+    else
+    {
+        w = sqrt(w) - 3.0;
+        p = -0.000200214257;
+        p = 0.000100950558 + p*w;
+        p = 0.00134934322 + p*w;
+        p = -0.00367342844 + p*w;
+        p = 0.00573950773 + p*w;
+        p = -0.0076224613 + p*w;
+        p = 0.00943887047 + p*w;
+        p = 1.00167406 + p*w;
+        p = 2.83297682 + p*w;
+    }
+    return p*x;
+}
+
+float lengthManhattan(float v)
+{
+    return abs(v);
+}
+float lengthManhattan(float2 v)
+{
+    v = abs(v);
+    return v.x + v.y;
+}
+float lengthManhattan(float3 v)
+{
+    v = abs(v);
+    return v.x + v.y + v.z;
+}
+float lengthManhattan(float4 v)
+{
+    v = abs(v);
+    return v.x + v.y + v.z + v.w;
+}
```

Comment:
could be templated if you replace the return with a `return dot(vector_t(1),v);`